LibreHardwareMonitorLib/Hardware/ISensor.cs

```diff
@@ -121,4 +122,15 @@ public interface ISensor : IElement
         /// </summary>
         void ResetMax();
     }
+
+    /// <summary>
+    /// Provides optional critical values for a sensor. This could be maximal temperatures, minimal voltages, etc.
+    /// </summary>
+    public interface ISensorCriticals
```

Comment:
This should be in a separate file. I think it should be ICriticalSensor to follow usual naming.
I'm also not entirely sure what I would expect for High and Low here. Shouldn't this derive from ISensor and then implement CriticalMin and CriticalMax?

Reply:
Deriving from `ISensor` would make inheriting sensor implementation hardly possible. CPUs have both maximal working and critical temperatures.

Reply:
Hmm, then I think it would make more sense to add another interface to specify those. Although I'm not sure how useful adding those are as as CPU should shut down at maximum temperature 😊

Reply:
Sorry, I do not understand what do you suggest. BTW, GPUs probably have both critical temps (throttling and shutdown) too.

Reply:
I mean to separate critical temperatures (i.e. throttle) and shutdown temperatures in two interfaces, I don't think all devices have both. Does that make sense? :)

Reply:
But then not all devices have low critical values, only high ones. Separate even further (just random thoughts)? In any case this is for you to maintain, so I will change it as you like.

Reply:
True, I'm just wondering what make sense to add. Is there a device that even specifies minimum temperatures? We can keep the interface as one and rename the properties to logical names and separate it later if needed. What about CriticalMin/Max and AbsoluteMin/Max? I'm open for other ideas 😊

Reply:
There are only minimal voltages and fan speeds, as far as I aware of PC hardware. 

Perhaps in the future there will be notifications for reaching critical values, and it would make sense to allow users to provide their own limits. Then I would think of something as follows:
```cs
enum Critical {
    Living,
    Surviving,
    User // unavailable for hardware sensors 
}

interface ISensorCriticals {
    Tuple<float,float> Range(Critical type);   // can return [-inf, 100.0] for temperatures, for example
}
```

Reply:
Or maybe like this:
```cs
enum SensorValueRange {
    Available,
    Optimal,
    Custom
}

interface ISensor {
...
Tuple<float, float> ValueRange(SensorValueRange rangeType);
}
```

`Available` makes sense for network interfaces too.

Reply:
I don't think that would make it much clearer than separate interfaces. I think it's best to keep Min and Max in one interface for now (and fallback to MinValue/MaxValue if it's unspecified).

Reply:
Tried this approach and I think the code became clearer.

Reply:
I don't think so. If needed in the future, it makes splitting up the code for sensors without a min or max not so logical.

Reply:
Do you mean those like CPU that only provide the high critical values? Do you want two methods instead of `ValueRange()`? Could you elaborate, please? I mean IEEE floats provide plenty of values to say "unlimited", why do you want to define that at compile time?

Reply:
I mean something like this:
```csharp
interface ICriticalSensor{
    float CriticalMin;
    float CriticalMax;
}

// Or another name that makes sense.
interface IAbsoluteSensor {
    float AbsoluteMin;
    float AbsoluteMax;
}
```

If it's then ever necessary it's easy to separate these in Min and Max versions through inheritance.

Reply:
Thank you, but I still don't understand. We already need two sets of critical values, how does the  `ICriticalSensor` interface from above can do that? Then, "absolute" is rather vague term, and in connection to temperature it has a very different meaning.

Reply:
// Or another name that makes sense. 😊

With your version you're bound to ranges, not every device will have them. 
The idea is to add a separate interface describing the sensor values as you need, similar to ISensor and my example above. So you'll have two additional interfaces. One with CriticalMin/Max, and one for the other. Maybe OptimalMin/Max.

Reply:
> The idea is to add a separate interface describing the sensor values as you need…

This is exactly what I can't understand. What advantages does it provide? All we need those ranges for is to compose tool tip text and likely to provide notifications when value exceed these ranges. For that we will have a single piece of code for each notification method, and not a code per sensor class. Therefore why would we need to learn which types of limits a sensor provides from its declaration? The only required information is does it provide limits or not, because any type of those are handled identically. Now, following this way the interface would look something like:
```csharp
interface ISensorCricticals {
    public enum Type {
        ...
    }

    float? Min(Type t);
    float? Max(Type t);
}
```

It would make sense if there was a way to provide default implementations for the `Min()` function, since, as you pointed out, most of the hardware have only high limits. But there is no because that is an interface and every inheritance need to provide explicit implementations. If so, there is no pros in two separate methods and we arrive to 
```
csharp
interface ISensorCricticals {
    public enum Type {
        ...
    }

    Tuple<float?, float?> Range(Type t);
}
```
but since every floating point type provides minus infinity and not-a-number, this can be further simplified replacing the nullable type with number.

Reply:
It's a clear separation from interface and implementation. You also know immediately which values you can expect from a sensor. With your structure you don't know what possible values a sensor has unless you loop through it. Maybe a user wants only sensors with critical values. Keep in mind that this is in the Lib and not just UI, other applications can (and likely will) use it too. 

A solution in the middle would look something like:
```csharp
interface ICriticalSensor<TSensorType> where TSensorType : SensorType
{
    float Min();
    float Max();
}
```
But this is not perfect either as you need an explicit interface implementation if you add more than one sensor type. I think the cleanest solution is multiple sensor interfaces. Then you have a clear distinction between different sensors and you also don't have to specify all available sensor types anymore.

BTW, since .NET Core 3.0 you can have default implementations in interfaces. This is of no use yet but perhaps in the future when .NET Framework gets dropped.

Reply:
> It's a clear separation from interface and implementation.

Er? Both approaches use interfaces, thus both separate interface from implementation. 

> You also know immediately which values you can expect from a sensor. 

How? I only know its working range is limited, and in your approach I also know exceeding certain value might lead to hardware failure or shutdown. But why do I need to know that at compile time?

> With your structure you don't know what possible values a sensor has unless you loop through it.

Same with yours because numeric values of the Min and Max properties are not part of the interface.

I've got a feeling you assume I want to add that method returning the value ranges to the `ISensor` interface, but I don't.

Reply:
> I've got a feeling you assume I want to add that method returning the value ranges to the `ISensor` interface, but I don't.

No, I understand what you mean. 

> Same with yours because numeric values of the Min and Max properties are not part of the interface.

You already have Min/Max from ISensor. But if you have a sensor instance, you can easily check ```if (sensor is ICriticalSensor)```, you know what it IS and HAS. If you just have ```ISensorCriticals``` you have to pass in all sensor types and see if anything returns some useful value. As I also mentioned, if we'd ever want to separate it into Min and Max interfaces (as not everything has a min and max), we can then easily do so with specific Min/Max interfaces (where the "normal" one can derive from). And even if it's not that useful right now in LHM UI, a lot of other applications are using LHM. They only consume the Lib.

I do hope you understand now the reasoning behind this 👍

Reply:
> if you just have `ISensorCriticals` you have to pass in all sensor types and see if anything returns some useful value.

No, it is not, because one has to test for `null`/NaN values for each property anyway.

Reply:
I mean you have to test anyway, with `sensor is ICriticalSensor` too, unless we create an interface for each property (for each critical/non-critical and min/max combination), but that would be crazy.

Reply:
> 
> 
> > if you just have `ISensorCriticals` you have to pass in all sensor types and see if anything returns some useful value.
> 
> No, it is not, because one has to test for `null`/NaN values for each property anyway.

You might have to check depending on what your interested in, but with your solution you don't know if a sensor implements critical values or not (or others) as it just returns defaults. And yes, you can return a variety of different values with different meanings, but that's not a nice solution. Anyway, I have made my point several times now, this is the last I've said about it 😊

Reply:
> but with your solution you don't know if a sensor implements critical values or not 

No, I [do](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/pull/458/files#diff-fc330d8dd912a4b72039528e76ae3fe0d556380a4a9bb7d62e59980871c8cf03R217). It is the same as with any other solution, as nothing prevents sensor class to implement a particular interface and return garbage values. 

> Anyway, I have made my point several times now, this is the last I've said about it 

I still don't understand how do you want to implement that, unfortunately. Can we somehow proceed? Maybe you could provide a code fragment, please? Specifically, for the [SensorNode.Tolltip](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/pull/458/files#diff-fc330d8dd912a4b72039528e76ae3fe0d556380a4a9bb7d62e59980871c8cf03R213) property?

Reply:
You only know it implements some kind of sensor value, but you don't know which one. If a sensor only returns garbage for an interface then it shouldn't implement it.

So basically add an interface per kind of sensor value. Something like this (Optimal (or something else) instead of Absolute might make more sense, the names are just given as-is, it depends on the kind of value):  https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/pull/458#discussion_r649876829
So one example would be (only Min/Max, as a current Value doesn't make sense here):
```csharp
public interface ICriticalSensor
{
    float? CriticalMin { get; }
    float? CriticalMax { get; }
}
```
The result would be that a class such as CorsairPsu could then derive from both these interfaces. So 
```csharp
public class CorsairPsu : ICriticalSensor, IOptimalSensor
``` 
and then implement the interfaces accordingly.
In tooltip you simply check 
```csharp
if (sensor is ICriticalSensor criticalSensor) 
    // add criticalSensor info to tooltip, such as toolTip += criticalSensor.CriticalMax?.ToString();
if (sensor is IOptimalSensor optimalSensor) 
    // add optimalSensor info to tooltip
```

Reply:
Thank you and take a look, please.

Reply:
Oh, no, It can't work this way or I don't know how to make it work: I need the general "Low + High limits" abstraction plus two specializations. One can't specialize generic interface with value it seems (only types are allowed), and inheritance does not work the way I want (seems like implementing an interface is always virtual inheritance in C++ terms). I don't know how to resolve this problem. Adding different names for various criticals is against generalization and leads to code duplication.

Reply:
You don't need that (although it's certainly possible if you use explicit interface implementations), they can be simply be two different types. They are different data after all, one is critical values, the other one is optimal values.

Reply:
The compiler says that the properties belong to the base interface if I try to add the derived name in the explicit implementation. It behaves as if the base interface was declared as virtual base in C++. 

Since we struggle with these criticals, I will extract the base part out of this PR and then can will refine the API for critical values later on, OK?

Reply:
All you need to do is split these up in two separate interfaces. Use different names, they have different meanings after all. So you'd have:

```csharp
public interface ICriticalSensorLimits
{
    float? MaxCriticalLimit { get; }
    float? MinCriticalLimit { get; }
}
```

and

```csharp
public interface ISensorLimits
{
    float? MaxLimit { get; }
    float? MinLimit { get; }
}
```

Reply:
BTW, you can read up on explict interface implementations here if you're interested (I don't find it suitable for sensors): https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces/explicit-interface-implementation.

Reply:
I'd rather went with PR splitting, because, firstly, we don't have good names for these public interfaces yet, and, secondly, I still hope to invent a more pleasant solution. As for the explicit interface implementations, I'm whining on the following:
```cs
interface IPropertyHolder {
     int prop { get; }
}

interface INamedProperty1: IPropertyHolder {}
interface INamedProperty2: IPropertyHolder {}

class C: INamedProperty1, INamedProperty2
{
    // there is only single IPropertyHolder.prop here, being explicitly implemented or not 
}
```

A generic interface `IPropertyHolder` would help, but then again one can't specialize generics with values in .NET.

Reply:
Sure, feel free to split it up. In your example above: in class C you would add INamedProperty**1**.prop and INamedProperty**2**.prop. Please take a look at the MSDN doc above. But I don't think that's a perfect solution as you would always need to cast it first, two interfaces seems like the way to go.

Reply:
Wait a minute, in the older C++ days we used to create tag types for this kind of situations. We can use that pattern here too:
```cs
namespace CriticalTags {
     public class Tag {};
     public sealed class Normal: Tag{};
     public sealed class Critical: Tag{};
}

interface ISensorCriticals<T>  where T: CriticalTags.Tag {
     float? Min { get; }
     float? Max { get; }
}


using INormalLimits = ISensorCriticals<CriticalTags.Normal>;
using ICriticallLimits = ISensorCriticals<CriticalTags.Critical>;

class Sensor: INormalLimits, ICriticallLimits
{
    // ...
}
```

Had not test with the compiler, but provided it works, what would you say about that?

Reply:
> In your example above: in class C you would add INamedProperty1.prop and INamedProperty2.prop.

No, the compiler says there are no such properties, only `IPropertyHolder.prop`, I tried that already.

Reply:
> Wait a minute, in the older C++ days we used to create tag types for this kind of situations.

You already proposed this yesterday :D

Reply:
I gave that as an example of a middle ground, but I also said it's not a perfect solution as you always need to cast it down to the interface. The better solution here is to add two interfaces. It's trivial to check for two interfaces in the tooltip and add the info it contains. We're talking about two "repeated" properties here (which isn't really repeated as they have totally different meanings). Compare to your example that adds a ton of additional and unnecessary clutter.

Reply:
> 
> 
> > In your example above: in class C you would add INamedProperty1.prop and INamedProperty2.prop.
> 
> No, the compiler says there are no such properties, only `IPropertyHolder.prop`, I tried that already.

You're doing something wrong then, it surely works: https://dotnetfiddle.net/wEZT3p

Reply:
Eh, nevermind that one, you of course mean values from devired interfaces.

Reply:
I give up :)

Reply:
@PhyxionNL, I meant it is now redone to your liking.

Reply:
@zeule I haven't forgotten about this, it's quite big so it'll take some time to review 👀

Reply:
Sure! Take your time, please.